authhelper.go

```diff
@@ -0,0 +1,176 @@
+	Email              string `json:"email,omitempty"`
+	ModelVersionNumber int    `json:"modelVersionNumber"`
+}
+
+const baseURL = "https://singlestore.com/engine-sso"
+
+func main() {
+	main2(os.Stdout)
+}
+
+func main2(stdout io.Writer) {
+	var config struct {
+		Email     string   `flag:"email" validate:"omitempty,validate" help:"users SSO email address, if known"`
+		ClusterID []string `flag:"cluster-id,split=comma" help:"comma-separated list of specific clusters to access"`
+		Databases []string `flag:"databases,split=comma" help:"comma-separated list of specific databases to access"`
+	}
+	flagFiller := nfigure.PosixFlagHandler(nfigure.WithHelpText(""))
+	reg := nfigure.NewRegistry(nfigure.WithFiller("flag", flagFiller))
+	err := reg.Request(&config)
+	if err != nil {
+		panic(err.Error())
+	}
+	err = reg.Configure()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	path := "/" + randomdata.Alphanumeric(20)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Using httptest since it takes care of picking a random port
+	var svr *httptest.Server
+	svr = httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		done := handle(w, r, svr, stdout, path)
+		if done {
+			wg.Done()
+		}
+	}))
+
+	// We do all of this to figure out when the request we're looking for has
+	// completed so that we can then exit the program.
+	activeConnections := make(map[string]struct{})
+	var lock sync.Mutex
+	svr.Config.ConnState = func(conn net.Conn, state http.ConnState) {
+		connStr := conn.RemoteAddr().Network() + conn.RemoteAddr().String()
+		lock.Lock()
+		defer lock.Unlock()
+		switch state {
+		case http.StateIdle, http.StateClosed, http.StateHijacked:
+			if _, ok := activeConnections[connStr]; ok {
+				wg.Done()
+				delete(activeConnections, connStr)
+			}
+		case http.StateActive, http.StateNew:
+			if _, ok := activeConnections[connStr]; !ok {
+				activeConnections[connStr] = struct{}{}
+				wg.Add(1)
+			}
+		}
+	}
+
+	svr.Start()
+
+	browser.Stdout = browser.Stderr
+
+	values := url.Values{}
+	values["returnTo"] = []string{svr.URL + path}
+	if config.Email != "" {
+		values["email"] = []string{config.Email}
+	}
+	if len(config.ClusterID) != 0 {
+		values["cluster"] = config.ClusterID
+	}
+	if len(config.Databases) != 0 {
+		values["db"] = config.Databases
+	}
+	url := baseURL + "?" + values.Encode()
+	browser.OpenURL(url)
+
+	wg.Wait()
+
+	svr.Close()
+}
+
+type Claims struct {
+	jwt.RegisteredClaims
+	Email      string `json:"email"`
+	DBUsername string `json:"dbUsername"`
+}
+
+func (c Claims) Valid() error {
+	err := c.RegisteredClaims.Valid()
+	if err != nil {
+		return err
+	}
+	if c.Email == "" {
+		return fmt.Errorf("Missing 'email' in claims")
+	}
+	if !c.VerifyExpiresAt(time.Now().Add(time.Minute), true) {
+		return fmt.Errorf("Invalid/missing 'exp' in claims")
+	}
+	return nil
+}
+
+func handle(w http.ResponseWriter, r *http.Request, svr *httptest.Server, stdout io.Writer, path string) bool {
```

Comment:
if I'm reading this right, all that needs to be passed as the body for this endpoint is the JWT string retrieved through the auth process right?

Reply:
Yes.